core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -2036,6 +2036,8 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
     dagScheduler.cancelStage(stageId)
   }
 
+  private val cleanerMap = new ConcurrentHashMap[Class[_], java.lang.Boolean]()
```

Comment:
Will a weak map be good here ? Especially for long running system , and a system which is predominantly job or UDF based , this will have hard references.